### PR TITLE
-Z call-metadata: add call metadata to LLVM IR

### PR DESCRIPTION
**Pull request by japaric**
This PR implements eRFC rust-lang/rust#59412 in spirit but there are some
differences between the implementation and the original proposal.

## What

Here's a summary of *what* the `-Z call-metadata` feature does:

1. Each function that:

- may be used as a function pointer,
- it's a method of a trait that may be dispatched dynamically,
- or may be called by trait object drop glue

receives a *single* `!rust` metadata node when translated into LLVM IR (turns
out you can't attach several `!rust` metadata nodes as I originally thought).
The node describes in which of those categories the function falls into (it may
belong to more than one category).

As an example consider this piece of code:

``` rust
fn foo() -> i32 {
    0
}

trait Bar<T> {
    fn bar(&self) -> T;
}

struct Baz;

impl Bar<bool> for Baz {
    fn bar(&self) -> bool {
        true
    }
}

impl Bar<i32> for Baz {
    fn bar(&self) -> i32 {
        0
    }
}

struct Quux;

impl Bar<i32> for Quux {
    fn bar(&self) -> i32 {
        1
    }
}

fn main() {
    let x: fn() -> i32 = foo;
    x();
    let y: fn(&Baz) -> bool = Baz::bar;

    let z: Box<dyn Bar<bool>> = Box::new(Baz);
    let a = z.bar();
    drop(z);
    let mut w: &dyn Bar<i32> = &Quux;
    w = &Baz;
    let b = w.bar();
}
```

Using the `-Z call-metadata` flag produces the following (unoptimized) IR:

> NOTE mangled names and debug metadata have been omitted to keep the IR short

``` llvm
; core::ptr::real_drop_in_place
define internal void @_(%Baz* nonnull align 1) unnamed_addr #4 !rust !180 {
  ; ..
}

; core::ptr::real_drop_in_place
define internal void @_(%Quux* nonnull align 1) unnamed_addr #4 !rust !190 {
  ; ..
}

; hello::foo
define internal i32 @_() unnamed_addr #4 !rust !363 {
  ; ..
}

; <hello::Baz as hello::Bar<bool>>::bar
define internal zeroext i1 @_(%Baz* noalias nonnull readonly align 1) unnamed_addr #4 !rust !370 {
  ; ..
}

; <hello::Baz as hello::Bar<i32>>::bar
define internal i32 @_(%Baz* noalias nonnull readonly align 1) unnamed_addr #4 !rust !377 {
  ; ..
}

; <hello::Quux as hello::Bar<i32>>::bar
define internal i32 @_(%Quux* noalias nonnull readonly align 1) unnamed_addr #4 !rust !377 {
  ; ..
}

!180 = !{!"drop", !"Bar<i32>", !"Bar<bool>"}
!190 = !{!"drop", !"Bar<i32>"}
!363 = !{!"fn", !"fn() -> i32"}
!370 = !{!"dyn", !"Bar<bool>", !"bar", !"fn", !"fn(&Baz) -> bool"}
!377 = !{!"dyn", !"Bar<i32>", !"bar"}
```

Functions that may be called via a function pointer receive metadata of the
form: `!"fn" !"fn() -> i32"`, where the second node is the signature of the
function. Example: `!363`

Trait methods that may be called via dynamic dispatch receive metadata of the
form: `!"dyn" !"Bar<bool>" !"bar"`, where the second node is a concrete trait
and the third node is the name of the method. Example: `!377`

Note that a single function can receive both `"fn"` and `"dyn"` metadata as
it's the case of `<Baz as Bar<i32>::bar` in the example above (`!370`).

Destructors that may be invoked by trait object drop glue receive metadata of
the form: `!"drop" !"Bar<i32>"`, where the second node is a concrete trait.
Example: `!180`

A single destructor may be invoked by different trait objects (because a type
can implement several traits); in that case the "drop" metadata will contain
one node for each trait, as it's the case of `Baz`'s destructor in the above
example (`!190`).

2. Indirect function calls will receive metadata at call site if they belong to
   any of these groups:

- Function pointer calls,
- Dynamic dispatch, or
- Trait object drop glue

Using our previous example, this is the part of the IR that shows the call site
metadata added by the `-Z call-metadata` flag:

``` llvm
; hello::main
define internal void @() unnamed_addr #0 {
  ; ..

  ; `x()`
  %1 = call i32 %0(), !rust !363

  ; ..

  ; `let a = z.bar();`
  %14 = invoke zeroext i1 %13({}* align 1 %8)
          to label %bb4 unwind label %cleanup, !db!rust !430

  ; ..

  ; `let b = w.bar();`
  %37 = invoke i32 %36({}* align 1 %31)
          to label %bb6 unwind label %cleanup, !rust !377

  ; ..

  ; `drop(z)`
  call void @_ZN4core3ptr18real_drop_in_place17h9dde6f5cdf273fbdE(..) #26

  ; ..
}

; this is `Box<dyn Bar<bool>>`'s destructor
; core::ptr::real_drop_in_place
define internal void @_ZN4core3ptr18real_drop_in_place17h9dde6f5cdf273fbdE(..) {
  ; ..

  ; `drop_in_place::<dyn Bar<bool>>(_)`
  invoke void %8({}* align 1 %3)
          to label %bb3 unwind label %cleanup, !rust !211

  ; ..
}

!211 = !{!"drop", !"Bar<bool>"}
!363 = !{!"fn", !"fn() -> i32"}
!377 = !{!"dyn", !"Bar<i32>", !"bar"}
!430 = !{!"dyn", !"Bar<bool>", !"bar"}
```

A function invocation may only belong to a single category. The syntax used for
call site metadata is as follows:

- `!"fn" "fn() -> i32"` for function pointer calls. The second node is the
  signature of the function being invoked. Example: `!363`

- `!"dyn" !"Bar<bool>" !"bar"` for dynamic dispatch. The second node is a
  concrete trait and the third node is the name of the method being dispatched.
  Examples: `!377` and `!430`

- `!"drop"` !"Bar<bool>" for drop glue. The second node is a concrete trait. In
  the case of call site "drop" metadata only a single trait will listed. Example: `!211`

## How

And here's a high level description of *how* the flag is implemented.

In the MIR monomorphize collector pass we take note of:

- Functions converted into function pointers. These occur in two places: in
  non-const context (e.g. `let x: fn() = foo`), and in const-context (`static X:
  fn() = foo`). These hit two different code paths in the compiler (while
  walking the MIR of functions and when looking into MIR values of static variables) but in
  both cases we keep track of which `Instance`s (`rustc` type used to track
  concrete functions) get converted into a function pointer in a `FxHashSet`.

- Traits that can get converted into trait objects. Again these occur in two
  places: non-const context (e.g. `let x: &dyn Foo = &Bar`) and in const-context
  (`static X: &'static dyn Foo = &Bar`). As in the previous bullet these hit two
  different code paths in the compiler but in both cases we keep track of *all
  methods (`Instance`s) that belong to these traits* in a `FxHashSet`.

- Drop glue of trait objects. While performing bullet 2 we also note down the
  types that implement traits that get converted into trait objects. In this
  case we store in the information in a `FxHashMap` that goes from the
  *implementer* destructor (the `Instance` e.g. `real_drop_inplace(&mut Bar)`)
  to the set of concrete traits (`ExistentialTraitRef`s) implemented by the
  implementer type (only the ones that get converted into trait objects).

(Now I realize that for the second bullet it would use less memory to keep
a set of `ExistentialTraitRef`s (traits) instead of a set of `Instance`s (trait
methods)).

Then in the LLVM codegen pass:

- When declaring functions (lowering them to `define` items in LLVM IR) we
  attach "define"-level metadata if the function (`Instance`) is in one of the
  sets / maps we previously collected. The functions get either "drop"
  metadata, a mixture of "dyn" and "fn" metadata or no `!rust` metadata.

- When lowering function invocations to `call` / `invoke` instructions we look up
  the callee (`Instance`) in the sets / maps we previously collected and if
  there's a match we add either "fn", "dyn" or "drop" metadata.

## TODO / questions

This doesn't fully work cross crate. Call site metadata is always complete, even
if the function being invoked comes from a different crate; and in the case of
LTO, invocations in functions defined in dependencies will also get call site
metadata. The "define"-level metadata, however, is incomplete; only functions
defined (or monomorphize) in the top crate get this kind of metadata.

To get complete "define"-level metadata I think we would have to store the call
metadata ("is this function converted into function pointer?", "is this trait
converted into a trait object?", etc.) in the rlib metadata (like we do for
types and traits). However, that would mean either (a) *always* building and
storing the call metadata, which would impact everyone's compile times and
increase rlib size, or (b) only storing the metadata when the `-Z call-metadata`
flag is used, which would worsen end-users' experience as they would need to
recompile `core` / `std` to get full information. If we do (b) we could compile
the `std` facade with `-Z call-metadata` so end users don't need to recompile it
themselves (which requires a tool like Xargo).

My gut feeling (and hope) is that always building and storing the extra maps /
sets won't have much impact on compile time given that it's the existing MIR
pass plus a few extra trait lookups and map/set insertions per indirect function
call. So perhaps we can always build and store the call metadata?

r? @eddyb or @oli-obk

### CODE REVIEW

File: src/librustc/query/mod.rs
Comment (oli-obk) [inline src/librustc/query/mod.rs]: This has collected quite the list of output arguments. I think it would be better to use a struct with named fields here

File: src/librustc/ty/instance.rs
Comment (oli-obk) [inline src/librustc/ty/instance.rs]: use the `?` operator here

File: src/librustc/ty/instance.rs
Comment (oli-obk) [inline src/librustc/ty/instance.rs]: Please document this method

File: src/librustc_mir/monomorphize/collector.rs
Comment (oli-obk) [inline src/librustc_mir/monomorphize/collector.rs]: I don't like how this file has been affected by the change. It causes some significant readability overhead that I feel is irrelevant to what the collector actually does. Implementation wise I think I'd be fine with it if you created a new type that has all the logic and datastructures in a single place (in another file preferrably). If there are just some calls of the `call_metadata.foo()` sort flying around this file that seems much better to me.

File: src/librustc_codegen_ssa/lib.rs
Comment (oli-obk) [inline src/librustc_codegen_ssa/lib.rs]: Document this type. It should be clear both here and at the use sites that this is solely for debug info and does not actually affect codegen

File: src/librustc_codegen_llvm/common.rs
Comment (oli-obk) [inline src/librustc_codegen_llvm/common.rs]: the `llvm::*` functions here should have safe wrappers on `CodegenCx` instead of requiring this code to use unsafe so freely.

**Review by oli-obk — CHANGES_REQUESTED** [review]
If you unconditionally enable the flag, we can test it out on perfbot. 

>  If we do (b) we could compile
the std facade with -Z call-metadata so end users don't need to recompile it
themselves (which requires a tool like Xargo).

We should try this out, both compilation speed of libstd as well as the dist size. Although I'm not sure what an ok regression would be

**Comment by eddyb** [comment]
We should probably also wait on a change like this: https://github.com/rust-lang/rust/issues/59412#issuecomment-480615364.

**Comment by japaric** [comment]
I have added the rustc_metadata encoding / decoding (it's still missing a small
bit) and refactored the implementation according to the review comments but now
I have realized that there's a fundamental problem with the approach of this PR. The
problem is easier to understand with an example so here's one:

Consider these two crates:

``` rust
// crate: a

pub fn foo() {
    // ..
}
```

``` rust
// crate: b

fn main() {
    let f: fn() = a::foo();

    // ..
}
```

In this example we want to attach "fn" metadata to `a::foo` the problem is that
we codegen that item when we codegen the crate `a` (and store the LLVM bitcode
in `liba.rlib` (\*)) but only learn about the "fn" metadata when we analyze
crate `b`.

(\*) For some reason I though that generating LLVM IR / bitcode only happened at
the very end when doing (fat) LTO but now I see that rustc always produces
(some) LLVM bitcode when it compiles a crate and fat LTO is just merging the
LLVM bitcode from all crates.

I'm a bit stumped now. Some ideas:

a. Is it possible to override the `define` metadata of items defined in other
  LLVM bitcode modules from a single LLVM bitcode module (i.e. from the top
  crate)?

b. Would it be feasible to add new `define` metadata after merging all LLVM
  bitcode modules into a single one when doing fat LTO? Is there LLVM API
  for bitcode manipulation?

c. Discard the idea of `define` metadata and instead emit all that information
  as some separate (JSON?) file. Meaning that tools would need to parse that
  file *and* the callsite metadata in the LLVM IR. This would be a shame because
  I think having all the metadata in the LLVM IR is pretty elegant and easy to
  work with.

d. Something else?

**Comment by oli-obk** [comment]
> d. Something else?

Am I understanding it correctly, that you'll need to know all the call metadata when codegening to llvm? If so, we could delay codegen to the final crate (so basically MIR-only rlibs).

**Comment by japaric** [comment]
@oli-obk that's correct.

> we could delay codegen to the final crate (so basically MIR-only rlibs).

*can* we implement that in this PR? Or do you mean to say that this is blocked on MIR-only rlibs?

**Comment by oli-obk** [comment]
> Or do you mean to say that this is blocked on MIR-only rlibs?

Only the "delay codegen to the final crate" solution is blocked on MIR-only rlibs. I don't remember what the blockers were. IIRC it was just that the codegen was slower or worse than with the current setup.

I don't know of another way to do this nicely.

**Comment by eddyb** [comment]
@japaric I think there's a flag you can use on nightly (`-Z always-encode-mir` maybe?) which has that behavior, but I could be wrong.

**Comment by oli-obk** [comment]
That flag will only encode the MIR for everything. It doesn't change the way llvm is invoked.

**Comment by japaric** [comment]
I have limited bandwidth right now so I'll come back to this PR at a later date.

**Pull request closed without merging**
